src/ScriptEditor/NetscriptDefinitions.d.ts

```diff
@@ -3914,11 +3919,44 @@ export interface CodingContract {
    * Note that this is not the same as the contract’s description.
    * This is just the data that the contract wants you to act on in order to solve the contract.
    *
+   * There is an optional third parameter where you can assert the contract type.
+   * This will type the data returned for that specific contract type.
+   * Be aware that this function will throw if the type in the fourth parameter is not the same as the type of the contract with that filename.
+   *
    * @param filename - Filename of the contract.
    * @param host - Host of the server containing the contract. Optional. Defaults to current server if not provided.
    * @returns The specified contract’s data, data type depends on contract type.
    */
-  getData(filename: string, host?: string): CodingContractData;
+  getData(filename: string, host?: string): any;
+  getData<T extends CodingContractName | `${CodingContractName}`>(
+    filename: string,
+    host: string | undefined,
+    type: T,
+  ): CodingContractData<T>;
+
+  /**
+   * Get the type and data of the contract.
+   * @remarks
+   * RAM cost: 10 GB
+   *
+   * Both the type and data of the contract are returned in an object.
+   * Depending on the type of the contract, the data is typed differently.
+   * Using type-narrowing, you can get the correct type of the data:
+   *
+   * @example
+   * ```js
+   * const contract = ns.codingcontract.getContract(fileName, hostName);
+   * if (contract.type === ns.enums.CodingContractName.FindLargestPrimeFactor) {
+   *   const data = contract.data;
+   *   // ^? data: number
+   * }
+   * ```
+   *
+   * @param filename - Filename of the contract.
+   * @param host - Host of the server containing the contract. Optional. Default to the current server if not provided.
+   * @returns An object containing both the type and data, typed depending on the type.
+   */
+  getContract(filename: string, host?: string): CodingContractSet;
```

Comment:
```suggestion
  getContract(filename: string, host?: string): CodingContractObj;
```
If the name "CodingContract" is already taken and can't be used, then tack on something generic. "Set" had collection connotations.